lib/internal/fs/cp/cp-sync.js

```diff
@@ -133,18 +133,22 @@ function setDestTimestamps(src, dest) {
 
 // TODO(@anonrig): Move this function to C++.
 function onDir(srcStat, destStat, src, dest, opts) {
-  if (!destStat) return mkDirAndCopy(srcStat.mode, src, dest, opts);
+  if (!destStat) return copyDir(src, dest, opts, true, srcStat.mode);
   return copyDir(src, dest, opts);
 }
 
-function mkDirAndCopy(srcMode, src, dest, opts) {
-  mkdirSync(dest);
-  copyDir(src, dest, opts);
-  return setDestMode(dest, srcMode);
-}
+function copyDir(src, dest, opts, mkDir, srcMode) {
+  if (!opts.filter) {
+    // the caller didn't provide a js filter function, in this case
+    // we can run the whole function faster in C++
+    // TODO(dario-piotrowicz): look into making cpSyncCopyDir also accept the potential filter function
```

Comment:
I can also look into the `filter` function in this PR if preferred but I figured it could make sense to do attempt that separately